artiq/gateware/drtio/transceiver/eem_serdes.py

```diff
@@ -0,0 +1,523 @@
+from migen import *
+from migen.genlib.resetsync import AsyncResetSynchronizer
+from migen.genlib.cdc import MultiReg
+from migen.genlib.misc import WaitTimer
+from misoc.interconnect.csr import *
+from misoc.cores.code_8b10b import SingleEncoder, Decoder
+from artiq.gateware.drtio.core import TransceiverInterface, ChannelInterface
+
+from operator import add
+
+
+class SerdesCRG(Module, AutoCSR):
+    def __init__(self, platform, main_clk, clk_freq, with_reset):
+        if with_reset:
+            self.mmcm_reset = CSRStorage(reset=1)
+            self.mmcm_locked = CSRStatus()
+        self.clock_domains.cd_eem_sys = ClockDomain()
+        self.clock_domains.cd_eem_sys5x = ClockDomain(reset_less=True)
+
+        eem_fb_in = Signal()
+        eem_fb_out = Signal()
+        mmcm_locked = Signal()
+        mmcm_eem_sys = Signal()
+        mmcm_eem_sys5x = Signal()
+
+        self.specials += [
+            Instance("MMCME2_BASE",
+                p_CLKIN1_PERIOD=1e9/clk_freq,
+                i_CLKIN1=main_clk,
+                i_RST=self.mmcm_reset.storage if with_reset else 0,
+
+                i_CLKFBIN=eem_fb_in,
+                o_CLKFBOUT=eem_fb_out,
+                o_LOCKED=mmcm_locked,
+
+                # VCO @ 1.25 GHz
+                p_CLKFBOUT_MULT_F=1.25e9/clk_freq, p_DIVCLK_DIVIDE=1,
+
+                # 125MHz
+                p_CLKOUT0_DIVIDE_F=10, p_CLKOUT0_PHASE=0.0, o_CLKOUT0=mmcm_eem_sys,
```

Comment:
In this case we can't keep both 4x clocks for the RAM. The 5x clock needs to be 0.8 times the divider of the 4x clocks, where the 4x dividers need to be integers because there are only 1 float divider.

Using the float divider looks fine, but it has a resolution of 0.125, so the least 4x divider is actually 5. VCO would be 2.5 GHz, exceeds the maximum of 1.6 GHz by far.

Reply:
There is also a phase 90.0 SDRAM clock, besides the phase 0 one.